runtime/lua/vim/lsp.lua

```diff
@@ -1612,95 +1612,20 @@ function lsp.buf_notify(bufnr, method, params)
 ---
 ---@see |complete-functions|
 ---@see |complete-items|
 ---@see |CompleteDone|
 ---
----@param findstart 0 or 1, decides behavior
----@param base If findstart=0, text to match against
+---@param findstart number 0 or 1, decides behavior
+---@param base number If findstart=0, text to match against
 ---
 ---@returns (number) Decided by {findstart}:
 --- - findstart=0: column where the completion starts, or -2 or -3
 --- - findstart=1: list of matches (actually just calls |complete()|)
 function lsp.omnifunc(findstart, base)
-  local _ = log.debug() and log.debug("omnifunc.findstart", { findstart = findstart, base = base })
-
-  local bufnr = resolve_bufnr()
-  local has_buffer_clients = not tbl_isempty(all_buffer_active_clients[bufnr] or {})
-  if not has_buffer_clients then
-    if findstart == 1 then
-      return -1
-    else
-      return {}
-    end
-  end
-
-  -- Then, perform standard completion request
-  local _ = log.info() and log.info("base ", base)
-
-  local pos = vim.api.nvim_win_get_cursor(0)
-  local line = vim.api.nvim_get_current_line()
-  local line_to_cursor = line:sub(1, pos[2])
-  local _ = log.trace() and log.trace("omnifunc.line", pos, line)
-
-  -- Get the start position of the current keyword
-  local textMatch = vim.fn.match(line_to_cursor, '\\k*$')
-
-  local params = util.make_position_params()
-
-  local items = {}
-  lsp.buf_request(bufnr, 'textDocument/completion', params, function(err, result, ctx)
-    if err or not result or vim.fn.mode() ~= "i" then return end
-
-    -- Completion response items may be relative to a position different than `textMatch`.
-    -- Concrete example, with sumneko/lua-language-server:
-    --
-    -- require('plenary.asy|
-    --         ▲       ▲   ▲
-    --         │       │   └── cursor_pos: 20
-    --         │       └────── textMatch: 17
-    --         └────────────── textEdit.range.start.character: 9
-    --                                 .newText = 'plenary.async'
-    --                  ^^^
-    --                  prefix (We'd remove everything not starting with `asy`,
-    --                  so we'd eliminate the `plenary.async` result
-    --
-    -- `adjust_start_col` is used to prefer the language server boundary.
-    --
-    local client = lsp.get_client_by_id(ctx.client_id)
-    local encoding = client and client.offset_encoding or 'utf-16'
-    local candidates = util.extract_completion_items(result)
-    local startbyte = adjust_start_col(pos[1], line, candidates, encoding) or textMatch
-    local prefix = line:sub(startbyte + 1, pos[2])
-    local matches = util.text_document_completion_list_to_complete_items(result, prefix)
-    -- TODO(ashkan): is this the best way to do this?
-    vim.list_extend(items, matches)
-    vim.fn.complete(startbyte + 1, items)
-  end)
-
-  -- Return -2 to signal that we should continue completion so that we can
-  -- async complete.
-  return -2
+  return require('vim.lsp.completion').omnifunc(findstart, base)
```

Comment:
This results in two "omnifunc" functions: `vim.lsp.omnifunc` and `vim.lsp.completion.omnifunc`. 

What did we gain here? The cost is now more API surface area and documentation clutter.

Reply:
Would you be okay if I moved the function and deprecated the old one? 

Part of the motivation was actually to reduce the API surface by having everything in one place instead of split across lsp.lua and util.lua. That makes it possible to work more easily with local functions. 

This is picking up on some earlier efforts:

- https://github.com/neovim/neovim/pull/12874
- https://github.com/neovim/neovim/pull/16205

And preparation to continue on some of the pending features:

- https://github.com/neovim/neovim/pull/16102
- https://github.com/neovim/neovim/pull/16974
- https://github.com/neovim/neovim/pull/16225
- https://github.com/neovim/neovim/pull/16974

My plan was to add one function that let's one generate a customised omnifunc to handle all of that and potentially also

- https://github.com/neovim/neovim/issues/13199

I was also torn on keeping the extract_completion_items function.  It's a trivial function and I'm tending to deprecate it even without replacement

Reply:
why don't we move stuff into vim.lsp rather than create another submodule? we don't need so many modules